File: boundaries/link/internal/infrastructure/repository/crud/mongo/dto/link.go
package dto

import (
	"net/url"
	"time"

	v1 "github.com/shortlink-org/shortlink/boundaries/link/internal/domain/link/v1"
)

var (
	errEmptyURL      = v1.NewInvalidInputError("mongo dto: URL cannot be empty")
	errEmptyHash     = v1.NewInvalidInputError("mongo dto: hash cannot be empty")
	errNilDomainLink = v1.NewInternalError("mongo dto: domain Link is nil")
)

// Link represents the Data Transfer Object for Link.
type Link struct {
	// URL
	Url url.URL `bson:"url" json:"url"`
	// Hash by URL + salt
	Hash string `bson:"hash" json:"hash"`
	// Describe of a link
	Describe string `bson:"describe" json:"describe"`

	// Created at
	CreatedAt time.Time `bson:"created_at" json:"created_at"`
	// Updated at
	UpdatedAt time.Time `bson:"updated_at" json:"updated_at"`
}

// ToDomain converts the DTO Link to the domain Link.
func (d *Link) ToDomain() (*v1.Link, error) {
	if d.Url.String() == "" {
		return nil, errEmptyURL
	}

	if d.Hash == "" {
		return nil, errEmptyHash
	}

	// Create a new domain.Link instance
	domainLink, err := v1.NewLinkBuilder().
		SetURL(d.Url.String()).
		SetDescribe(d.Describe).
		SetCreatedAt(d.CreatedAt).
		SetUpdatedAt(d.UpdatedAt).
		Build()
	if err != nil {
		return nil, err
	}

	return domainLink, nil
}

// FromDomain converts the domain Link to the DTO Link.
func FromDomain(d *v1.Link) (*Link, error) {
	if d == nil {
		return nil, errNilDomainLink
	}

	return &Link{
		Url:       *d.GetUrl().URL,
		Hash:      d.GetHash(),
		Describe:  d.GetDescribe(),
		CreatedAt: d.GetCreatedAt().GetTime(),
		UpdatedAt: d.GetUpdatedAt().GetTime(),
	}, nil
}

File: boundaries/proxy/src/proxy/domain/entities/Link.ts
import { Hash } from "./Hash.js";

/**
 * Доменная сущность ссылки
 * Содержит бизнес-логику и инварианты домена
 */
export class Link {
  constructor(
    public readonly hash: Hash,
    public readonly url: string,
    public readonly createdAt: Date = new Date(),
    public readonly updatedAt: Date = new Date()
  ) {
    this.validate();
  }

  /**
   * Валидация инвариантов домена
   */
  private validate(): void {
    if (!this.url || typeof this.url !== "string" || this.url.trim().length === 0) {
      throw new Error("Link URL cannot be empty");
    }

    // Валидация URL формата (базовая проверка)
    try {
      new URL(this.url);
    } catch {
      throw new Error(`Invalid URL format: "${this.url}"`);
    }
  }

  /**
   * Проверяет, является ли ссылка валидной для редиректа
   */
  isValidForRedirect(): boolean {
    return this.url.length > 0 && this.hash.value.length > 0;
  }

  /**
   * Сравнивает две ссылки на равенство по хешу
   */
  equals(other: Link): boolean {
    return this.hash.equals(other.hash);
  }
}

